content/en/blog/_posts/2018-08-02-dynamically-expand-volume-csi.md

```diff
@@ -0,0 +1,163 @@
+
+
+```
+type ExpandableVolumePlugin interface {
+VolumePlugin
+ExpandVolumeDevice(spec Spec, newSize resource.Quantity, oldSize resource.Quantity) (resource.Quantity, error)
+RequiresFSResize() bool
+}
+```
+
+### Implement Volume Driver
+
+Finally, to abstract complexity of the implementation, we should write hard code to separate storage provider management logic into the following functions which is well-defined in the CSI specification:
+
+* CreateVolume
+* DeleteVolume
+* ControllerPublishVolume
+* ControllerUnpublishVolume
+* ValidateVolumeCapabilities
+* ListVolumes
+* GetCapacity
+* ControllerGetCapabilities
+* RequiresFSResize
+* ControllerResizeVolume
+
+## Demonstration
+
+Let’s demonstrate this feature with concrete user case.
+
+* Create storage class for CSI storage provisioner:
+
+```
+allowVolumeExpansion: true
+apiVersion: storage.k8s.io/v1
+kind: StorageClass
+metadata:
+  name: csi-qcfs
+parameters:
+  csiProvisionerSecretName: orain-test
+  csiProvisionerSecretNamespace: default
+provisioner: csi-qcfsplugin
+reclaimPolicy: Delete
+volumeBindingMode: Immediate
+```
+
+* Deploy CSI Volume Driver including storage provisioner `csi-qcfsplugin` across Kubernetes cluster.
+
+* Create PVC `qcfs-pvc` which will be dynamically provisioned by storage class `csi-qcfs`:
+
+```
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: qcfs-pvc
+  namespace: default
+....
+spec:
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 300Gi
+  storageClassName: csi-qcfs
+```
+
+* Create MySQL 5.7 instance to use PVC `qcfs-pvc`
+* In order to mirror the exact same production-level scenario, there are actually two different types of workloads including:
+    * Batch insert to make MySQL consuming more file system capacity;
+    * Surge query request;
+* Dynamically expand volume capacity through edit pvc `qcfs-pvc` configuration;
+
+The Prometheus and Grafana integration let us visualize corresponding critical metrics.
+
+![prometheus grafana](/images/blog/2018-08-02-dynamically-expand-volume-csi/prometheus-grafana.png)
+
+We could notice that the middle reading shows MySQL datafile size increasing slowly during bulk inserting. At the same time, the bottom reading shows file system expanding twice in about 20 minutes, from 300 GiB to 400 GiB and then 500 GiB. Meanwhile, the upper reading shows the whole process of expanding volume immediately completes and hardly impacts MySQL QPS.
+
+## Conclusion
+
+Regardless of whatever infrastructure applications have been running on, the database is always a critical resource. It is essential to have a more advanced storage subsystem out there to fully support database requirements.
+
+Furthermore, it is very critical to have more conservative companies realize that their monolithic applications which have been running for decades could be running properly on Kubernetes. At least in my opinion, it will be the real drivers for broad adoption of cloud native technology.
```

Comment:
> could also run properly on Kubernetes.

changed times here a little.

> At least in my opinion, it will be the real drivers for broad adoption of cloud native technology.

i'm unsure what `it` refers to.
i would move this sentence after the previous paragraph: 
> fully support database requirements.

to be come:
> fully support database requirements. This will help drive the more broad adoption of cloud native technology.